### PR TITLE
Adds as_core_arg serde module to network

### LINKED ISSUE #1799: Add `as_core_arg` module to implemenet `serde` using `Network::to_core_arg` and `Netwrok::from_core_arg`
This is useful in json RPC but others might find it useful too.

### PR DESCRIPTION
**Pull request by sr-gi**
The new module provides serde functions to convert from/to Bitcoin Core network variants.

Fixes #1799

### CODE REVIEW

**Review by tcharding — DISMISSED** [review]
Thanks for your contribution! Could we have a unit test to show this in action? Perhaps something like:
```rust

    #[cfg(feature = "serde")]
    #[test]
    fn serde_as_core_arg() {
        #[derive(Serialize, Deserialize, PartialEq, Debug)]
        #[serde(crate = "actual_serde")]
        struct T {
            #[serde(with = "crate::network::constants::as_core_arg")]
            pub network: Network,
        }

        serde_test::assert_tokens(
            &T { network: Network::Bitcoin },
            &[
                serde_test::Token::Struct { name: "T", len: 1 },
                serde_test::Token::Str("network"),
                serde_test::Token::Str("main"),
            ],
        );
    }
```

And to all other reviewers I think we should leave the path to `as_core_arg` as it is in this PR so as not to bother @sr-gi with bikeshedding (since there are multiple discussions currently ongoing in regards to module paths and exports).

File: bitcoin/src/network/constants.rs [hunk 1]
```diff
@@ -173,6 +173,44 @@ impl Network {
     }
 }
 
+#[cfg(feature = "serde")]
+pub mod as_core_arg {
```
Comment (Kixunil) [inline bitcoin/src/network/constants.rs:177]: Hmm, now that I think about it maybe we should name it `network_as_core_arg`. Anyway I also kinda dislike `Network` being in `constants`, so let's leave it for now and maybe move/rename later.

**Review by Kixunil — APPROVED** [review]
ACK 142373b434fdca30568b66bdf4085d7b2262ab3a

File: bitcoin/src/network/constants.rs [hunk 2]
```diff
@@ -173,6 +173,44 @@ impl Network {
...
+        impl<'de> serde::de::Visitor<'de> for NetworkVisitor {
+            type Value = Network;
+    
+            fn visit_str<E: serde::de::Error>(self, s: &str) -> Result<Self::Value, E> {
+                Network::from_core_arg(s)
+                    .map_err(|_| E::invalid_value(serde::de::Unexpected::Str(s), &"bitcoin network encoded as a string"))
+            }
+    
+            fn expecting(&self, formatter: &mut core::fmt::Formatter) -> core::fmt::Result {
+                write!(formatter, "bitcoin network encoded as a string")
```
Comment (Kixunil) [inline bitcoin/src/network/constants.rs:206]: Just realized, it might be nice to enumerate all possible networks in the message.

**Review by Kixunil — APPROVED** [review]
ACK 0d5e7a2c545eac74733fc01b567f0b28dbe417cf

**Comment by Kixunil** [comment]
Oh, CI wants you to run `cargo +nightly fmt`.

File: bitcoin/src/network/constants.rs (hunk 2 above)
Comment (sr-gi) [reply bitcoin/src/network/constants.rs]: Done in [4e9ae25](https://github.com/rust-bitcoin/rust-bitcoin/pull/1801/commits/4e9ae25626944e42ba0d54cb61de2e9918113e2f)

Comment (sr-gi) [reply bitcoin/src/network/constants.rs]: Done in aa6a3530a626c17a875c0d961ece2ef8ab73d7fb

**Review by Kixunil — APPROVED** [review]
ACK aa6a3530a626c17a875c0d961ece2ef8ab73d7fb

**Review by tcharding — APPROVED** [review]
Thanks man, awesome.

ACK aa6a3530a626c17a875c0d961ece2ef8ab73d7fb

**Comment by tcharding** [comment]
Oh, @sr-gi can you not use mentions in commits in future please, it causes github to spam the notifications a bunch of times. Thanks

**Comment by sr-gi** [comment]
> Oh, @sr-gi can you not use mentions in the commit log please, it causes github to spam the notifications a bunch of times. Thanks

I'll have that in mind in the future, sorry 😅

**Comment by tcharding** [comment]
No sweat man, I only learned it myself recently :)

**Review by sanket1729 — APPROVED** [review]
ACK aa6a3530a626c17a875c0d961ece2ef8ab73d7fb

**Pull request merged**